pkg/controllers/updaterun/initialization.go

```diff
@@ -0,0 +1,452 @@
+						return fmt.Errorf("%w: %s", errInitializedFailed, keyErr.Error())
+					}
+				}
+				curStageClusters = append(curStageClusters, cluster)
+				allPlacedClusters[cluster.Name] = struct{}{}
+			}
+		}
+
+		// Check if the stage is empty.
+		if len(curStageClusters) == 0 {
+			// since we allow no selected bindings, a stage can be empty.
+			klog.InfoS("No cluster is selected for the stage", "clusterStagedUpdateStrategy", updateStrategyName, "stage name", stage.Name, "clusterStagedUpdateRun", updateRunRef)
+		} else {
+			// Sort the clusters in the stage based on the SortingLabelKey and cluster name.
+			sort.Slice(curStageClusters, func(i, j int) bool {
+				if stage.SortingLabelKey == nil {
+					return curStageClusters[i].Name < curStageClusters[j].Name
+				}
+				labelI, _ := strconv.Atoi(curStageClusters[i].Labels[*stage.SortingLabelKey])
+				labelJ, _ := strconv.Atoi(curStageClusters[j].Labels[*stage.SortingLabelKey])
+				if labelI != labelJ {
+					return labelI < labelJ
+				}
+				return curStageClusters[i].Name < curStageClusters[j].Name
+			})
+		}
+
+		// Record the clusters in the stage.
+		curStageUpdatingStatus.Clusters = make([]placementv1alpha1.ClusterUpdatingStatus, len(curStageClusters))
+		for i, cluster := range curStageClusters {
+			klog.V(2).InfoS("Adding a cluster to the stage", "cluster", cluster.Name, "clusterStagedUpdateStrategy", updateStrategyName, "stage name", stage.Name, "clusterStagedUpdateRun", updateRunRef)
+			curStageUpdatingStatus.Clusters[i].ClusterName = cluster.Name
+		}
+
+		// Create the after stage tasks.
+		curStageUpdatingStatus.AfterStageTaskStatus = make([]placementv1alpha1.AfterStageTaskStatus, len(stage.AfterStageTasks))
+		for i, task := range stage.AfterStageTasks {
+			curStageUpdatingStatus.AfterStageTaskStatus[i].Type = task.Type
+			if task.Type == placementv1alpha1.AfterStageTaskTypeApproval {
+				curStageUpdatingStatus.AfterStageTaskStatus[i].ApprovalRequestName = fmt.Sprintf(placementv1alpha1.ApprovalTaskNameFmt, updateRun.Name, stage.Name)
+			}
+		}
+		updateRun.Status.StagesStatus = append(updateRun.Status.StagesStatus, curStageUpdatingStatus)
+	}
+
+	// Check if the clusters are all placed.
+	if len(allPlacedClusters) != len(allSelectedClusters) {
+		missingErr := fmt.Errorf("some clusters are not placed in any stage")
+		for cluster := range allSelectedClusters {
+			if _, ok := allPlacedClusters[cluster]; !ok {
+				klog.ErrorS(missingErr, "Cluster is missing in any stage", "cluster", cluster, "clusterStagedUpdateStrategy", updateStrategyName, "clusterStagedUpdateRun", updateRunRef)
+			}
+		}
+		// no more retries here.
+		return fmt.Errorf("%w: %s", errInitializedFailed, missingErr.Error())
+	}
+	return nil
+}
+
+// validateAfterStageTask valides the afterStageTasks in the stage defined in the clusterStagedUpdateStrategy.
+// The error returned from this function is not retryable.
+func validateAfterStageTask(tasks []placementv1alpha1.AfterStageTask) error {
```

Comment:
this probably can be converted to CEL validation

Reply:
Got an error when trying to add CEL validation: 
`unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD "clusterstagedupdateruns.placement.kubernetes-fleet.io": CustomResourceDefinition.apiextensions.k8s.io "clusterstagedupdateruns.placement.kubernetes-fleet.io" is invalid: spec.validation.openAPIV3Schema.properties[status].properties[stagedUpdateStrategySnapshot].properties[stages].items.properties[afterStageTasks].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic`